File: csharp/AdventOfCode19/AdventOfCode19/Day1.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AdventOfCode19
{
    public class Day1
    {
        public static int PartOne()
        {
            var masses = ParseData();
            int result = 0;
            foreach (int mass in masses)
            {
                result += GetFuelForMass(mass);                
            }
            return result;
        }

        public static int PartTwo()
        {
            var masses = ParseData();
            int result = 0;
            foreach (int mass in masses)
            {
                int currentFuel = GetFuelForMass(mass);
                int totalFuel = 0;
                while(currentFuel > 0)
                {
                    totalFuel += currentFuel;
                    currentFuel = GetFuelForMass(currentFuel);
                }
                result += totalFuel;
            }
            return result;
        }

        static int GetFuelForMass(int mass)
        {
            return (mass / 3) - 2;
        }        

        static List<int> ParseData()
        {
            var result = new List<int>();
            try
            {
                string[] lines = System.IO.File.ReadAllLines(@"..\..\..\..\..\data\1.txt");
                foreach (string line in lines)
                {
                    int? number = ConvertStringToInt(line);
                    if (number != null) result.Add((int)number);
                }
            }
            catch(System.IO.DirectoryNotFoundException err)
            {
                Console.WriteLine("Day 1 directory not found");
            }            
            return result;
        }
        static int? ConvertStringToInt(string intString)
        {
            int i = 0;
            return (Int32.TryParse(intString, out i) ? i : (int?)null);
        }
    }
}

File: rust/src/day1.rs
use std::fs::File;
use std::io::Read;

fn main() { 
}

pub fn part_one() -> i32{
  let filename = "../data/1.txt";
  let masses = parse_data(filename);
  let mut result = 0;
  for mass in masses {
    result += get_fuel_for_mass(mass);
  }
  result
}

pub fn part_two() -> i32{
  let filename = "../data/1.txt";
  let masses = parse_data(filename);
  let mut result = 0;
  for mass in masses {
    let mut current_fuel = get_fuel_for_mass(mass);
    let mut total_fuel = 0;
    while current_fuel > 0 {
      total_fuel += current_fuel;
      current_fuel = get_fuel_for_mass(current_fuel);
    }
    result += total_fuel;
  }
  result
}

fn get_fuel_for_mass(mass: i32) -> i32 {
  return (mass / 3) - 2;
}

fn parse_data(filename: &str) -> Vec<i32> {
  let mut file = match File::open(filename) {
      Ok(file) => file,
      Err(_) => panic!("no such file"),
  };
  let mut file_contents = String::new();
  file.read_to_string(&mut file_contents)
      .ok()
      .expect("failed to read!");
  let lines: Vec<i32> = file_contents.split("\n")
      .map(|s: &str| s.to_string())
      .filter(|s| !s.is_empty())
      .map(|s| s.parse().unwrap())
      .collect();
  lines
}